### PR TITLE
Prepare target models before running attacks

### PR DESCRIPTION
**Pull request by mzweilin**
# What does this PR do?

This PR adds two preparations before running attacks in an external Lightning pipeline.

1. Turn off the PyTorch inference mode, so that we can create perturbation variables that require gradients.
2. Switch the target model to the training mode except for BatchNorm and Dropout layers, if we have to borrow `training_step()`.

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] `pytest`
- [x] `CUDA_VISIBLE_DEVICES=0 python -m mart experiment=CIFAR10_CNN_Adv trainer=gpu trainer.precision=16` reports 70% (21 sec/epoch).
- [ ] `CUDA_VISIBLE_DEVICES=0,1 python -m mart experiment=CIFAR10_CNN_Adv trainer=ddp trainer.precision=16 trainer.devices=2 model.optimizer.lr=0.2 trainer.max_steps=2925 datamodule.ims_per_batch=256 datamodule.world_size=2` reports 70% (14 sec/epoch).

## Before submitting

- [x] The title is **self-explanatory** and the description **concisely** explains the PR
- [x] My **PR does only one thing**, instead of bundling different changes together
- [ ] I list all the **breaking changes** introduced by this pull request
- [x] I have commented my code
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have run pre-commit hooks with `pre-commit run -a` command without errors

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: mart/attack/adversary.py [hunk 1]
```diff
@@ -151,6 +151,8 @@ def configure_gradient_clipping(
             for group in optimizer.param_groups:
                 self.gradient_modifier(group["params"])
 
+    # Turn off the inference mode, so we will create perturbation that requires gradient.
+    @torch.inference_mode(False)
```
Comment (dxoigmn) [inline mart/attack/adversary.py:155]: Why is this necessary now? I thought PL manages this already?

File: mart/callbacks/adversary_connector.py [hunk 2]
```diff
@@ -7,15 +7,51 @@
...
+    def __init__(self, module, excludes=[]):
+        self.module = module
+        self.excludes = excludes
+
+    def __enter__(self):
+        # Save the original training mode status.
+        self.training = self.module.training
+        self.module.train(True)
+        # Set some children modules of "excludes" to eval mode instead.
+        self.selective_eval_mode("", self.module, self.excludes)
```
Comment (dxoigmn) [inline mart/callbacks/adversary_connector.py:36]: What is going on with the empty string?

File: mart/callbacks/adversary_connector.py [hunk 3]
```diff
@@ -76,12 +112,18 @@ def model(input, target):
...
                 outputs = pl_module.attack_step(batch, dataloader_idx)
             else:
-                # LightningModule must have "training_step".
-                # Disable logging if we have to reuse training_step() of the target model.
+                # If there is no attack_step(), we will borrow LightningModule:training_step() with some modifications.
+                #   1. Disable the training logging mechanism.
+                #   2. Switch to the training mode to get the loss value, except for children modules of BatchNorm and Dropout.
                 with MonkeyPatch(pl_module, "log", lambda *args, **kwargs: None):
-                    outputs = pl_module.training_step(batch, dataloader_idx)
+                    with training_mode(
```
Comment (dxoigmn) [inline mart/callbacks/adversary_connector.py:122]: What is the use case here? Are you seeing train-specific code diverging from eval-specific code in some use case?

File: mart/attack/adversary.py (hunk 1 above)
Comment (mzweilin) [reply mart/attack/adversary.py:155]: Anomalib turns on the inference mode as we run `anomalib test`.

MART's trainer turns off the inference mode by default, as in https://github.com/IntelLabs/MART/blob/05886f7606872f69480eebdb8ee0435c8b873b48/mart/configs/trainer/default.yaml#L19

But Anomalib has its own trainer.

File: mart/callbacks/adversary_connector.py (hunk 2 above)
Comment (mzweilin) [reply mart/callbacks/adversary_connector.py:36]: We don't know the variable name of the model, so the module path starts with a dot. This is for debug logging only, which print messages like this 

```
Set .model.student_model.feature_extractor.layer3[1].bn1: BatchNorm2d to eval mode.
```

File: mart/callbacks/adversary_connector.py (hunk 3 above)
Comment (mzweilin) [reply mart/callbacks/adversary_connector.py:122]: Yes. Many model implementations return the prediction in the eval mode, and return the loss in the training mode.

In our use case, `anomalib test` runs the model in the eval mode, in which we won't get the loss.

**Review by dxoigmn — APPROVED** [review]
